### PR TITLE
Show error message when JS code fails to load

### PR DESCRIPTION
**Pull request by Turbo87**
This should resolve https://github.com/rust-lang/crates.io/issues/2984, by adding a short error message when something fails to load before the Ember app has booted up:

Bildschirmfoto 2020-11-25 um 00 31 31

This does use an inline script though, so we might have to adjust the CSP for this 😞 

r? @jtgeibel

### CODE REVIEW

**Comment by Turbo87** [comment]
I've added a hash of the inline script to the CSP config for nginx, deployed it to staging, and it appears to work :)

File: app/index.html [hunk 1]
```diff
@@ -28,6 +28,7 @@
 
     {{content-for 'body'}}
 
+    <script>window.onerror=function(){document.body.innerHTML='<p style="width: 70%;background: var(--main-bg);padding: 10px;">Sorry, it looks like we were not able to load the page. Please make sure your network connection works and you are using an up-to-date browser. If the issue persists, please visit our <a href="https://github.com/rust-lang/crates.io/issues/new/choose">issue tracker</a> to report the problem.</p>'}</script>
```
Comment (jtgeibel) [inline app/index.html:31]: I think we should add a comment above this line, stating that any modifications to the inline script require an update to the CSP hash.

**Review by jtgeibel — CHANGES_REQUESTED** [review]
This works for me locally in Firefox and Chrome when I block the `vendor-*.js` resource, but this does not work (i.e. blank page only) if I block `cargo-*.js`.

Additionally, in Firefox only, I see the error `None of the “sha512” hashes in the integrity attribute match the content of the subresource.` in my console. This is confusing because I don't have CSP enabled in my local environment, and the error seems to be referring to [Subresource Integrity](https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity). Otherwise the behavior between Chrome and Firefox appears the same, so this may be unrelated.

I don't understand why this appears to rely on cargo.js loading.

My test setup is to run `ember build --prod` followed by `cargo run --bin server` and then accessing via `http://localhost:8888`. This is similar to our production enviornment, sans nginx configuration (like CORS).

**Comment by Turbo87** [comment]
I've managed to reproduce your issue and I think I understand what's going on:

At least in Chrome, it looks like blocking a URL through the devtools does not cause the `onerror` handler to be called, instead the request is just silently ignored. The reason why blocking the `vendor.js` file "worked" is that the `cargo.js` file depends on code in the `vendor.js` file, and since that code wasn't there it caused an "Uncaught ReferenceError: define is not defined" error visible on the devtools console.

In other words: blocking the URL through the devtools does not seem to create a realistic failure scenario. The thing we want to test here is having a syntax error in the JS file, so what I did to test this is prefix the content of one of the JS files with e.g. an `@` sign so that the browser can't parse it as valid JS anymore. This worked for both `cargo.js` and `vendor.js`, so I assume that this is still working as intended.

File: app/index.html (hunk 1 above)
Comment (Turbo87) [reply app/index.html:32]: added :)

**Comment by jtgeibel** [comment]
Well it sounds like my test steps were fatally flawed then. Thanks for investigating further and adding the comment.

@bors r+
